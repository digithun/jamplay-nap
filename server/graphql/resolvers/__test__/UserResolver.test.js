/* eslint-env jest */
const _NOW_PLUS_60SEC_ISO = new Date(+new Date() + 1000 * 60).toISOString()

describe('UserResolver', () => {
  it('should return null if session not provided', async () => {
    const { user: willReadUser } = require('../UserResolver')
    const context = { nap: { errors: [] } }
    const user = await willReadUser({ context })
    expect(user).toBeNull()
  })

  it('should return user data if has session', async () => {
    // mock
    const userData = { foo: 'bar' }

    // stub
    global.NAP = {}
    NAP.User = {
      findById: jest.fn().mockImplementationOnce(() =>
        Promise.resolve(
          Object.assign(
            {
              _id: '592c0bb4484d740e0e73798b',
              role: 'user'
            },
            userData
          )
        )
      )
    }

    const { user: willReadUser } = require('../UserResolver')
    const context = { nap: { session: { userId: 'foo', expireAt: _NOW_PLUS_60SEC_ISO } } }
    const user = await willReadUser({ context })

    expect(user).toMatchSnapshot()
  })

  it('should able to forget password', async () => {
    const context = {
      headers: { host: 'localhost:3000' },
      nap: {
        willResetPasswordViaEmail: async () => ({
          status: 'WAIT_FOR_EMAIL_RESET'
        })
      }
    }
    const email = 'foo@bar.com'
    const args = { email }
    const { forget } = require('../UserResolver')
    const user = await forget({ context, args })
    expect(user).toMatchSnapshot()
  })

  it('should able to update email', async () => {
    const context = {
      nap: {
        session: {
          userId: '592c0bb4484d740e0e73798b',
          expireAt: _NOW_PLUS_60SEC_ISO
        },
        willUpdateEmail: require('../../../../server/authen-local-passport').willUpdateEmail
      }
    }
    const args = { email: 'foo@bar.com' }

    // stub
    global.NAP = {}
    NAP.User = {
      findById: jest.fn().mockImplementationOnce(async () => ({
        _id: '592c0bb4484d740e0e73798b',
        save: async () => ({
          _id: '592c0bb4484d740e0e73798b'
        })
      })),
      findOne: jest.fn().mockImplementationOnce(async () => null)
    }

    const { updateEmail } = require('../UserResolver')
    const user = await updateEmail({ context, args })
    delete user.save
    expect(user).toMatchSnapshot()
  })

  it('should not able to update existing email', async () => {
    const context = {
      nap: {
        session: {
          userId: '592c0bb4484d740e0e73798b',
          expireAt: _NOW_PLUS_60SEC_ISO
        },
        willUpdateEmail: require('../../../../server/authen-local-passport').willUpdateEmail,
        errors: []
      }
    }
    const args = { email: 'foo@bar.com' }

    // stub
    global.NAP = {}
    NAP.User = {
      findById: jest.fn().mockImplementationOnce(async () => ({
        _id: '592c0bb4484d740e0e73798b',
        save: async () => ({
          _id: '592c0bb4484d740e0e73798b'
        })
      })),
      findOne: jest.fn().mockImplementationOnce(async () => ({
        _id: '592c0bb4484d740e0e73798c'
      }))
    }

    const { updateEmail } = require('../UserResolver')
    expect(updateEmail({ context, args })).rejects.toMatchObject(require('../../../errors/codes').AUTH_EMAIL_ALREADY_IN_USE)
  })
})
