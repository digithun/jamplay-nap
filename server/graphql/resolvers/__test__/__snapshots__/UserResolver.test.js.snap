// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserResolver should able to forget password 1`] = `
Object {
  "status": "WAIT_FOR_EMAIL_RESET",
}
`;

exports[`UserResolver should able to update email 1`] = `
Object {
  "_id": "592c0bb4484d740e0e73798b",
}
`;

exports[`UserResolver should create user and return user data 1`] = `
Object {
  "_id": "592c0bb4484d740e0e73798b",
  "foo": "bar",
  "role": "user",
}
`;

exports[`UserResolver should return error if session not exist 1`] = `[Error: A uid identifier is required for the current operation.]`;

exports[`UserResolver should return user data if has session 1`] = `
Object {
  "_id": "592c0bb4484d740e0e73798b",
  "foo": "bar",
  "role": "user",
}
`;
